### PR TITLE
39 teacher profile

### PR DESCRIPTION
**Pull request by AlaaTaima**
Relates #39

### CODE REVIEW

File: server/controllers/routes/teacherProfile.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+const { getTeacherProfile } = require('../../database/quieres/teacherProfile');
+
+require('env2')('./config.env');
+
+const secret = process.env.SECRET_KEY;
+
+const teacherProfile = (req, res, next) => {
+  getTeacherProfile()
+    .then(({ rows }) => {
+      if (req.cookies.access) {
```
Comment (11janci) [inline server/controllers/routes/teacherProfile.js]: why are you getting around the `auth` middleware?
Here you first get the data from the DB and then verify the cookie - that's wrong, should be the other way around. And with this approach you would have to repeat this code in every API that doesn't use the `auth` middleware, so in the very least it should have been extracted into a function, but using the middleware is the correct approach, this doesn't look right.

File: server/database/quieres/teacherProfile.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const connection = require('../config/connection');
+
+const getTeacherProfile = () => connection.query(
+  'select * from teacher inner join "user" on teacher.user_id="user".id ',
```
Comment (11janci) [inline server/database/quieres/teacherProfile.js]: selecting everything from the user table with return the users' passwords! that's a serious security flaw, despite the fact the passwords are hashed :) I'll create an issue for this, but here you can just select only the columns you actually need

File: server/controllers/routes/teacherProfile.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        jwt.verify(req.cookies.access, secret, (error, value) => {
+          if (value) {
+            res.json(rows);
+          } else {
+            res.redirect('/');
+          }
+        });
+      } else res.redirect('/');
```
Comment (11janci) [inline server/controllers/routes/teacherProfile.js]: again, redirect; `else` should be on a new line, or do
```
} else {
    ...
}
```

File: server/controllers/routes/teacherProfile.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+const teacherProfile = (req, res, next) => {
+  getTeacherProfile()
+    .then(({ rows }) => {
+      if (req.cookies.access) {
+        jwt.verify(req.cookies.access, secret, (error, value) => {
+          if (value) {
+            res.json(rows);
+          } else {
+            res.redirect('/');
```
Comment (11janci) [inline server/controllers/routes/teacherProfile.js]: why redirect? This is API call, not the browser..

Comment (AlaaTaima) [reply server/controllers/routes/teacherProfile.js]: thanks, I got you and I edited it

File: server/controllers/routes/teacherProfile.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const { getTeacherProfile } = require('../../database/quieres/teacherProfile');
+
+
+const teacherProfile = (req, res, next) => {
+  getTeacherProfile()
```
Comment (11janci) [inline server/controllers/routes/teacherProfile.js]: this query currently selects all of the teachers in the database. If that's the goal, you should rename the API and the query to (for example) `getTeachersList`. 
If however the goal of this API is to retrieve one specific teacher profile (as the current name suggests), it needs to be parametrized with the ID of the profile to retrieve:
```
const { id } = req.params;
getTeacherProfile(id).then(...)
```

File: server/database/quieres/teacherProfile.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const connection = require('../config/connection');
+
+const getTeacherProfile = () => connection.query(
+  'select first_name, last_name, bio, image_url from teacher inner join "user" on teacher.user_id="user".id ',
```
Comment (11janci) [inline server/database/quieres/teacherProfile.js]: related to my previous comment, if this is supposed to return just one specific user, you need to parametrize the query
```
... on teacher.user_id="user".id WHERE "user".id = $1',
// pass the id as the parameter
```

File: server/router/index.js [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
-const { logout, login } = require('../controllers');
-
 const { auth } = require('../controllers/middlewares');
+const { logout, login, teacherProfile } = require('../controllers');
 
 const router = express.Router();
 
 router.post('/login', login);
 router.use(auth);
+router.get('/profile/teacher', teacherProfile);
```
Comment (11janci) [inline server/router/index.js]: and here, again if this is supposed to search for a specific profile, you'd need
```
router.get('/profiles/teachers/:id', teacherProfile);
```

File: server/controllers/middlewares/index.js [hunk 8]
```diff
@@ -1,4 +1,4 @@
 const { auth } = require('./auth');
+const { teacherProfile } = require('../routes/teacherProfile');
```
Comment (11janci) [inline server/controllers/middlewares/index.js]: don't think this should be here, it's not a middleware :) it's (correctly) exported through `controllers/index.js`

File: server/controllers/routes/teacherProfile.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    .then(({ rows }) => {
+      res.json(rows);
+    })
+    .catch((err) => next(err));
```
[the remark this answers is not in the document]
Comment (AlaaTaima) [reply server/controllers/routes/teacherProfile.js]: ok , I will edit it

File: server/controllers/routes/teacherProfile.js (hunk 1 above)
Comment (AlaaTaima) [reply server/controllers/routes/teacherProfile.js]: I got you , and I have edit the code correctly

File: server/database/quieres/teacherProfile.js (hunk 6 above)
Comment (AlaaTaima) [reply server/database/quieres/teacherProfile.js]: Done

**Pull request merged**